### ISSUE TITLE
[Data] Stratification in train_test_split

### ISSUE THREAD

**Issue by wingkitlee0** [ray-project/ray#53297]
### Description

To implement `stratify` option in `Dataset.train_test_split`

Currently, the non-stratified split is implemented random shuffle + split by index. The stratified version may require a groupby + train_test_split + merge back into train/test sets. It will be very great if these can be lazy.

scikit-learn:
* https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html
* https://scikit-learn.org/stable/modules/cross_validation.html#stratification

### Use case

The use cases are similar to `groupby.sample` in https://github.com/ray-project/ray/issues/53296

There are cases we cannot train-test-split uniformly if the group distribution matters (e.g., imbalanced dataset)

**Comment by krishnakalyan3** [comment]
I would like to work on this. Is there an example PR that I can use as a reference. Many Thanks.

**Comment by wingkitlee0** [comment]
@krishnakalyan3 this one may be a little more involved than typical "good first issue". Happy to discuss the idea first here or on Ray slack first before implementation.
For previous PR you can find some merged ones with `data` label.

**Comment by krishnakalyan3** [comment]
Thanks for your feedback @wingkitlee0. This would be my approach.

Approach:
1. Add stratify parameter to existing [train_test_split](https://github.com/ray-project/ray/blob/master/python/ray/data/dataset.py#L2253)() 
2. Use block-level operations to maintain laziness
3. Leverage existing GroupedData infrastructure
4. Shuffle the dataset
5. Return the final train, test split

Related:
https://github.com/ray-project/ray/issues/34849

**Comment by wingkitlee0** [comment]
ya...

similar to the previous suggestion, I was thinking something similar but couldnt think of something memory efficient for large dataset..
```
(
  ds.groupby(stratify_column)
  .map_groups(
   train_test_split_per_group, # also add a new column "is_train"
  )
  .repartition(keys="is_train")  # or other ways to split
)
```

I guess you could still to implement something and ask Ray team for review..

**Comment by wingkitlee0** [comment]
@krishnakalyan3 Good job on https://github.com/ray-project/ray/pull/54624 

Closing this issue.
